File: osu-beatmap-preview-rust/src/service.rs
use crate::errors::{PreviewError, Result};
use crate::models::{Beatmap, HitObjects};
use crate::mods::{mods_for_mode, validate_mods, ModSettings};
use serde_json::{json, Map, Value};
use std::path::PathBuf;

pub fn generate_preview(
    bid: &str,
    fmt: Option<&str>,
    convert: Option<&str>,
    mods: Option<ModSettings>,
    times: Option<Vec<f64>>,
) -> Result<Value> {
    if bid.is_empty() || !bid.chars().all(|c| c.is_ascii_digit()) {
        return Err(PreviewError::new("bid must be numeric"));
    }

    let temp_root = std::env::temp_dir().join("osu-beatmap-preview");
    let beatmap_path =
        crate::downloader::download_beatmap_file(bid, &temp_root.join("osu-download-cache"))?;
    let beatmap = crate::parser::parse_beatmap(&beatmap_path)?;

    let mut target_mode = beatmap.mode();
    if let Some(convert_name) = convert {
        target_mode = crate::convert::resolve_convert_target(&beatmap, convert_name)?;
    }

    let fmt: String = match fmt {
        Some(f) => f.to_string(),
        None => {
            if target_mode == 0 {
                "gif".to_string()
            } else {
                "png".to_string()
            }
        }
    };
    if times.is_some() && fmt != "gif" {
        return Err(PreviewError::new("--times is only valid for GIF output"));
    }

    let mods: Option<ModSettings> = match mods {
        Some(m) if m.has_any_mod() => {
            let mode_errors = validate_mods(&m, Some(target_mode), Some(&fmt));
            if !mode_errors.is_empty() {
                return Err(PreviewError::new(format!(
                    "mod conflict: {}",
                    mode_errors.join("; ")
                )));
            }
            Some(mods_for_mode(&m, target_mode))
        }
        _ => None,
    };

    let mut parts: Vec<String> = vec![bid.to_string()];
    if let Some(convert_name) = convert {
        parts.push(convert_name.to_string());
    }
    if let Some(m) = &mods {
        if m.has_any_mod() {
            parts.push(format_mod_suffix(m));
        }
    }
    let output_path: PathBuf = temp_root
        .join("outputs")
        .join(format!("{}.{}", parts.join("_"), fmt));

    let preview_path =
        render_preview_for_mode(beatmap.clone(), &output_path, &fmt, target_mode, mods, times)?;

    let abs = preview_path
        .canonicalize()
        .unwrap_or(preview_path.clone());
    let abs_str = clean_windows_path(&abs.to_string_lossy());

    Ok(json!({
        "status": "success",
        "msg": format!("preview generated successfully for bid {bid}"),
        "preview-img": abs_str,
        "beatmap-info": {
            "meta-data": format_section_keys(&beatmap.metadata),
            "difficulty": format_section_keys(&beatmap.difficulty),
        },
    }))
}

fn clean_windows_path(path: &str) -> String {
    path.strip_prefix(r"\\?\").unwrap_or(path).to_string()
}

fn format_section_keys(section: &crate::models::KvSection) -> Value {
    let mut map = Map::new();
    for (key, value) in &section.entries {
        map.insert(kebab_case(key), Value::String(value.clone()));
    }
    Value::Object(map)
}

fn kebab_case(key: &str) -> String {
    // pass 1: ([a-z0-9])([A-Z]) -> \1-\2 ; pass 2: ([A-Z]+)([A-Z][a-z]) -> \1-\2
    let chars: Vec<char> = key.chars().collect();
    let mut pass1 = String::with_capacity(key.len() + 4);
    for i in 0..chars.len() {
        pass1.push(chars[i]);
        if i + 1 < chars.len()
            && (chars[i].is_ascii_lowercase() || chars[i].is_ascii_digit())
            && chars[i + 1].is_ascii_uppercase()
        {
            pass1.push('-');
        }
    }
    let chars: Vec<char> = pass1.chars().collect();
    let mut pass2 = String::with_capacity(pass1.len() + 4);
    let mut i = 0;
    while i < chars.len() {
        pass2.push(chars[i]);
        // boundary between a run of uppercase and [A-Z][a-z]
        if chars[i].is_ascii_uppercase()
            && i + 2 < chars.len()
            && chars[i + 1].is_ascii_uppercase()
            && chars[i + 2].is_ascii_lowercase()
        {
            pass2.push('-');
        }
        i += 1;
    }
    pass2.to_lowercase()
}

fn format_mod_suffix(mods: &ModSettings) -> String {
    let tokens: Vec<String> = mods
        .tokens
        .iter()
        .map(|t| t.trim().to_lowercase())
        .filter(|t| !t.is_empty())
        .map(|t| {
            t.chars()
                .filter(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || *c == '.' || *c == '-')
                .collect::<String>()
        })
        .filter(|t| !t.is_empty())
        .collect();
    if tokens.is_empty() {
        return "mod".to_string();
    }
    format!("mod-{}", tokens.join("-"))
}

fn render_preview_for_mode(
    beatmap: Beatmap,
    output_path: &std::path::Path,
    fmt: &str,
    target_mode: i32,
    mods: Option<ModSettings>,
    times: Option<Vec<f64>>,
) -> Result<PathBuf> {
    let times_ms = crate::time_selection::times_to_milliseconds(times.as_deref());
    let mods_ref = mods.as_ref();

    match target_mode {
        0 => {
            let has_objects = matches!(&beatmap.hit_objects, HitObjects::Standard(v) if !v.is_empty());
            if !has_objects {
                return Err(PreviewError::new("standard beatmap has no hit objects"));
            }
            if fmt == "gif" {
                crate::standard::render_standard_gif(&beatmap, mods_ref, times_ms, output_path)?;
            } else {
                let image = crate::standard::render_standard_png(&beatmap, mods_ref)?;
                crate::composer::save_png(&image, output_path)?;
            }
            Ok(output_path.to_path_buf())
        }
        1 => {
            let beatmap = if beatmap.mode() != 1 {
                crate::convert::convert_beatmap(&beatmap, target_mode, mods_ref)?
            } else {
                beatmap
            };
            if fmt == "gif" {
                crate::taiko::render_taiko_gif(&beatmap, mods_ref, times_ms, output_path)?;
                Ok(output_path.to_path_buf())
            } else {
                crate::taiko::render_taiko_grid(&beatmap, output_path, mods_ref)
            }
        }
        2 => {
            let beatmap = if beatmap.mode() != 2 {
                crate::convert::convert_beatmap(&beatmap, target_mode, mods_ref)?
            } else {
                beatmap
            };
            if fmt == "gif" {
                crate::catch::render_catch_gif(&beatmap, mods_ref, times_ms, output_path)?;
                Ok(output_path.to_path_buf())
            } else {
                crate::catch::render_catch_grid(&beatmap, output_path, mods_ref)
            }
        }
        3 => {
            let beatmap = if beatmap.mode() != 3 {
                crate::convert::convert_beatmap(&beatmap, target_mode, mods_ref)?
            } else {
                beatmap
            };
            if fmt == "gif" {
                crate::mania::render_mania_gif(&beatmap, mods_ref, times_ms, output_path)?;
                Ok(output_path.to_path_buf())
            } else {
                crate::mania::render_mania_grid(&beatmap, output_path, mods_ref)
            }
        }
        _ => Err(PreviewError::new(format!(
            "unsupported beatmap mode: {target_mode}"
        ))),
    }
}

File: beatmap_preview/service.py
from __future__ import annotations

import re
import tempfile
from pathlib import Path

from .composer import save_animated_gif, save_png
from .downloader import download_beatmap_file
from .errors import PreviewError
from .models import Beatmap
from .mods import ModSettings, mods_for_mode, validate_mods
from .parser import parse_beatmap
from .standard.renderer import render_standard
from .taiko.renderer import render_taiko_grid
from .catch.renderer import render_catch_grid
from .mania.renderer import render_mania_grid


def generate_preview(
    bid: str,
    fmt: str | None = None,
    convert: str | None = None,
    mods: ModSettings | None = None,
    times: list[float] | None = None,
) -> dict[str, object]:
    if not bid.isdigit():
        raise PreviewError("bid must be numeric")

    temp_root = Path(tempfile.gettempdir()) / "osu-beatmap-preview"
    beatmap_path = download_beatmap_file(bid=bid, temp_dir=temp_root / "osu-download-cache")
    beatmap = parse_beatmap(beatmap_path)

    # ── 模式转换 ──
    target_mode = beatmap.mode
    if convert:
        from .convert import resolve_convert_target

        target_mode = resolve_convert_target(beatmap, convert)

    # ── 解析 fmt 默认值 ──
    if fmt is None:
        fmt = "gif" if target_mode == 0 else "png"
    if times is not None and fmt != "gif":
        raise PreviewError("--times is only valid for GIF output")

    # ── mod 校验 ──
    if mods is not None and mods.has_any_mod():
        mode_errors = validate_mods(mods, target_mode, fmt)
        if mode_errors:
            raise PreviewError("mod conflict: " + "; ".join(mode_errors))
        mods = mods_for_mode(mods, target_mode)
    else:
        mods = None

    parts = [bid]
    if convert:
        parts.append(convert)
    if mods is not None and mods.has_any_mod():
        parts.append(_format_mod_suffix(mods))
    output_path = temp_root / "outputs" / f"{'_'.join(parts)}.{fmt}"

    preview_path = _render_preview_for_mode(
        beatmap, output_path, fmt, target_mode, mods, times
    )

    return {
        "status": "success",
        "msg": f"preview generated successfully for bid {bid}",
        "preview-img": str(preview_path.resolve()),
        "beatmap-info": {
            "meta-data": _format_section_keys(beatmap.metadata),
            "difficulty": _format_section_keys(beatmap.difficulty),
        },
    }


def _format_section_keys(section: dict[str, str]) -> dict[str, str]:
    return {
        re.sub(
            r"([A-Z]+)([A-Z][a-z])", r"\1-\2",
            re.sub(r"([a-z0-9])([A-Z])", r"\1-\2", key),
        ).lower(): value
        for key, value in section.items()
    }


def _format_mod_suffix(mods: ModSettings) -> str:
    tokens = [token.strip().lower() for token in mods.tokens if token.strip()]
    if not tokens:
        return "mod"
    return "mod-" + "-".join(re.sub(r"[^a-z0-9.-]+", "", token) for token in tokens if token)


def _render_preview_for_mode(
    beatmap: Beatmap,
    output_path: Path,
    fmt: str,
    target_mode: int,
    mods: ModSettings | None,
    times: list[float] | None,
) -> Path:
    if target_mode == 0:
        from .models import StandardHitObject
        hit_objects = [ho for ho in beatmap.hit_objects if isinstance(ho, StandardHitObject)]
        if not hit_objects:
            raise PreviewError("standard beatmap has no hit objects")
        result = render_standard(beatmap, hit_objects, fmt, mods=mods, times=times)
        if fmt == "gif":
            frames, frame_duration_ms, loop = result
            save_animated_gif(frames, output_path, frame_duration_ms, loop)
        else:
            save_png(result, output_path)
        return output_path

    if target_mode == 1:
        if beatmap.mode != 1:
            from .convert import convert_beatmap
            beatmap = convert_beatmap(beatmap, target_mode, mods)
        if fmt == "gif":
            from .taiko.gif_renderer import render_taiko_gif

            frames, frame_duration_ms, loop = render_taiko_gif(beatmap, mods=mods, times=times)
            save_animated_gif(frames, output_path, frame_duration_ms, loop)
            return output_path
        return render_taiko_grid(beatmap, output_path, mods=mods)

    if target_mode == 2:
        if beatmap.mode != 2:
            from .convert import convert_beatmap
            beatmap = convert_beatmap(beatmap, target_mode, mods)
        if fmt == "gif":
            from .catch.gif_renderer import render_catch_gif

            frames, frame_duration_ms, loop = render_catch_gif(beatmap, mods=mods, times=times)
            save_animated_gif(frames, output_path, frame_duration_ms, loop)
            return output_path
        return render_catch_grid(beatmap, output_path, mods=mods)

    if target_mode == 3:
        if beatmap.mode != 3:
            from .convert import convert_beatmap
            beatmap = convert_beatmap(beatmap, target_mode, mods)
        if fmt == "gif":
            from .mania.gif_renderer import render_mania_gif

            frames, frame_duration_ms, loop = render_mania_gif(beatmap, mods=mods, times=times)
            save_animated_gif(frames, output_path, frame_duration_ms, loop)
            return output_path
        return render_mania_grid(beatmap, output_path, mods=mods)

    raise PreviewError(f"unsupported beatmap mode: {target_mode}")
